File: homework/assignment7/collcount.pl
% collcount.pl
% Spencer Baysinger
% 2024-04-29

% For CS 331 Spring 2024

% Assignment 7 - Exercise D
%   Write a Prolog source file collcount.pl as follows. Be sure to follow the Coding Standards.
%   Your file should define the predicate collcount/2; that is, collcount is a predicate that takes 2 arguments.
%   Predicate collcount should take arguments as follows: collcount(+n, ?c); that is, the first argument is an input, while the second is either an input or an output.
%   The idea is that n is a positive integer, and c is the number of iterations of the Collatz function required to take n to 1 (just as in the Forth exercise).
%   Your code does not need to do any type checking or other error checking.

% Note to self: Open SWI-PROLOG "swipl-win.exe" in the folder, and then type "['collcount.pl]." to load, then operate functions.

collcount(1, 0).
collcount(N, C) :-
    N > 1,
    (
        0 is N mod 2 ->
        M is N // 2
    ;
        M is N * 3 + 1
    ),
    collcount(M, C1),
    C is C1 + 1.

File: homework/assignment7/collcount.fs
\ collcount.fs
\ Spencer Baysinger
\ 2024-04-29

\ For CS 331 Spring 2024

\ Assignment 7 Exercise A - Running a Nilgai Program
  \ Secret message: The Magic Words are Squeamish Ossifrage

\ Assignment 7 - Exercise B: Programming in Forth
  \ Write a Forth source file collcount.fs as follows. Be sure to follow the Coding Standards.
  \ Your file should define the word collcount.
  \ Word collcount should have stack effect (n -- c), where n is a positive integer, and c is the number of iterations of the Collatz function (see Assignment 5) required to take n to 1.
  \ Your code does not need to do any type checking or other error checking

  \ Note to self: Run using gforth on a unix machine

: collcount ( n -- c )
  0 swap
  begin
    dup 1 >
  while
    dup 2 mod if
      3 * 1+
    else
      2 /
    then
    swap 1+ swap
  repeat
  drop ;